librad/src/git/refs.rs

```diff
@@ -50,68 +49,106 @@ use crate::{
 pub use crate::identities::git::Urn;
 pub use git_ext::Oid;
 
-/// The transitive tracking graph, up to 3 degrees
+/// The depth of the tracking graph (ie. [`Remotes`]) to retain per peer.
+// TODO(kim): bubble up as parameter
+pub const TRACKING_GRAPH_DEPTH: usize = 3;
+
+/// The transitive tracking graph.
+// **NOTE**: A recursion limit of 128 is imposed by `serde_json` when deserialising.
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-pub struct Remotes<A: PartialEq + Eq + Ord>(BTreeMap<A, BTreeMap<A, BTreeMap<A, ()>>>);
+pub struct Remotes<A: Ord>(BTreeMap<A, Box<Remotes<A>>>);
+
+impl<A: Ord> Deref for Remotes<A> {
+    type Target = BTreeMap<A, Box<Remotes<A>>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
 
-impl<A> Remotes<A>
-where
-    A: PartialEq + Eq + Ord + Hash,
-{
-    pub fn cutoff(self) -> BTreeMap<A, BTreeMap<A, ()>>
+impl<A: Ord> DerefMut for Remotes<A> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<A: Ord> FromIterator<A> for Remotes<A> {
+    fn from_iter<T>(iter: T) -> Self
     where
-        A: Clone,
+        T: IntoIterator<Item = A>,
     {
-        self.0
-            .into_iter()
-            .map(|(k, v)| (k, v.keys().map(|x| (x.clone(), ())).collect()))
-            .collect()
+        Self(
+            iter.into_iter()
+                .map(|a| (a, Box::new(Self::new())))
+                .collect(),
+        )
     }
+}
 
-    pub fn flatten(&self) -> impl Iterator<Item = &A> {
-        self.0.iter().flat_map(|(k, v)| {
-            iter::once(k).chain(
-                v.iter()
-                    .flat_map(|(k1, v1)| iter::once(k1).chain(v1.keys())),
-            )
-        })
+impl<A: Ord> Remotes<A> {
+    pub fn new() -> Self {
+        Self(BTreeMap::new())
     }
 
-    pub fn from_map(map: BTreeMap<A, BTreeMap<A, BTreeMap<A, ()>>>) -> Self {
-        Self(map)
-    }
+    /// Build a new [`Self`] with at most `depth` levels.
+    pub fn cutoff(self, depth: usize) -> Self {
+        if depth == 0 {
+            return Self(BTreeMap::default());
+        }
 
-    pub fn boxed(self) -> Box<Self> {
-        Box::new(self)
+        Self(self.0.into_iter().fold(BTreeMap::new(), |mut acc, (k, v)| {
+            acc.insert(k, Box::new(v.cutoff(depth - 1)));
+            acc
+        }))
     }
-}
 
-impl<A> Deref for Remotes<A>
-where
-    A: PartialEq + Eq + Ord + Hash,
-{
-    type Target = BTreeMap<A, BTreeMap<A, BTreeMap<A, ()>>>;
+    /// Modify [`Self`] to contain at most `depth` levels.
+    pub fn cutoff_mut(&mut self, depth: usize) {
+        if depth > 0 {
+            let depth = depth - 1;
+            for v in self.0.values_mut() {
+                v.cutoff_mut(depth)
+            }
+        } else {
+            self.clear()
+        }
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+    /// Yield all `A` values in an unspecified order.
```

Comment:
It would be good to note that they may not be unique as well. I didn't think of that until I saw the test below :)